src/components/Swap/Summary/index.tsx

```diff
@@ -27,59 +30,176 @@ const PriceImpactText = styled.span`
   color: ${({ theme }) => theme.error};
 `
 
+const PermitTooltipBody = styled.div`
+  max-width: 220px;
+`
+
+function PermitTooltipText({ text, content }: { text: ReactNode; content: ReactNode }) {
+  return (
+    <TooltipText placement="bottom" offset={10} text={text}>
+      <PermitTooltipBody>
+        <ThemedText.Caption>{content}</ThemedText.Caption>
+      </PermitTooltipBody>
+    </TooltipText>
+  )
+}
+
+function getAllowanceFailedAction(isApproved: boolean, retry: () => void): Action {
+  return {
+    message: isApproved ? (
+      <PermitTooltipText
+        text={t`Token approval failed`}
+        content={t`A signature is needed to trade this token on the Uniswap protocol. For security, signatures expire after 30 days.`}
+      />
+    ) : (
+      <PermitTooltipText
+        text={t`Permit approval failed`}
+        content={t`Permit2 allows safe sharing and management of token approvals across different smart contracts.`}
+      />
+    ),
+    onClick: retry,
+    color: 'warning',
+    children: <Trans>Try again</Trans>,
+  }
+}
+
+function getAllowancePendingAction(isApproved: boolean, cancel: () => void): Action {
+  return {
+    message: isApproved ? (
+      <PermitTooltipText
+        text={t`Approve token for trading`}
+        content={t`Gives you the ability to trade this token on the Uniswap protocol. For security, this will expire in 30 days.`}
+      />
+    ) : (
+      <PermitTooltipText
+        text={t`Approve permit`}
+        content={t`Permit2 allows safe sharing and management of token approvals across different smart contracts.`}
+      />
+    ),
+    icon: Spinner,
+    onClick: cancel,
+    children: <Trans>Cancel</Trans>,
+  }
+}
+
+enum ReviewState {
+  REVIEWING,
+  ALLOWANCE_PENDING,
+  ALLOWANCE_FAILED,
+  TRADE_CHANGED,
+  PENDING_SWAP,
+}
+
 function ConfirmButton({
   trade,
   onConfirm,
   onAcknowledgeNewTrade,
+  allowance,
 }: {
   trade: InterfaceTrade
   onConfirm: () => Promise<void>
   onAcknowledgeNewTrade: () => void
+  allowance: Allowance
 }) {
+  const [reviewState, setReviewState] = useState(ReviewState.REVIEWING)
+
   const { onSwapPriceUpdateAck, onSubmitSwapClick } = useAtomValue(swapEventHandlersAtom)
   const [ackTrade, setAckTrade] = useState(trade)
   const doesTradeDiffer = useMemo(
     () => Boolean(trade && ackTrade && tradeMeaningfullyDiffers(trade, ackTrade)),
     [ackTrade, trade]
   )
+  useEffect(() => {
+    if (doesTradeDiffer && reviewState === ReviewState.REVIEWING) {
+      setReviewState(ReviewState.TRADE_CHANGED)
+    } else if (!doesTradeDiffer && reviewState === ReviewState.TRADE_CHANGED) {
+      setReviewState(ReviewState.REVIEWING)
+    }
```

Comment:
🤔 feels like moving back to the `REVIEWING` state should be part of the callback for trade acked instead of in this useEffect

Reply:
This is to catch a situation where the trade changes twice, where the first change causes tradeMeaningfullyDiffers to be true, but then the second change brings it back into a range where tradeMeaningfullyDiffers is false (acknowledgement no longer needed)